core/tests/window/test_window.py

```diff
@@ -584,6 +584,129 @@ def test_close_rejected_handler(window, app):
     on_close_handler.assert_called_once_with(window)
 
 
+def test_on_gain_focus(window):
+    assert window._on_gain_focus._raw is None
+
+    on_gain_focus_handler = Mock()
+    window.on_gain_focus = on_gain_focus_handler
+
+    assert window.on_gain_focus._raw == on_gain_focus_handler
+
+    window._impl.simulate_on_gain_focus()
+
+    on_gain_focus_handler.assert_called_once_with(window)
+
+
+def test_on_lose_focus(window):
+    assert window.on_lose_focus._raw is None
+
+    on_lose_focus_handler = Mock()
+    window.on_lose_focus = on_lose_focus_handler
+
+    assert window.on_lose_focus._raw == on_lose_focus_handler
+
+    window._impl.simulate_on_lose_focus()
+
+    on_lose_focus_handler.assert_called_once_with(window)
+
+
+def test_on_show(window):
+    assert window.on_show._raw is None
+
+    on_show_handler = Mock()
+    window.on_show = on_show_handler
+
+    assert window.on_show._raw == on_show_handler
+
+    window._impl.simulate_on_show()
+
+    on_show_handler.assert_called_once_with(window)
+
+
+def test_on_hide(window):
+    assert window.on_hide._raw is None
+
+    on_hide_handler = Mock()
+    window.on_hide = on_hide_handler
+
+    assert window.on_hide._raw == on_hide_handler
+
+    window._impl.simulate_on_hide()
+
+    on_hide_handler.assert_called_once_with(window)
+
+
+def test_focus_events(app):
+    """The window can trigger on_gain_focus() and on_lose_focus()
+    event handlers, when the window gains or loses input focus."""
+    window1 = toga.Window()
+    window1.show()
+    window1_on_gain_focus_handler = Mock()
+    window1_on_lose_focus_handler = Mock()
+    window1.on_gain_focus = window1_on_gain_focus_handler
+    window1.on_lose_focus = window1_on_lose_focus_handler
+
+    window2 = toga.Window()
+    window2.show()
+    window2_on_gain_focus_handler = Mock()
+    window2_on_lose_focus_handler = Mock()
+    window2.on_gain_focus = window2_on_gain_focus_handler
+    window2.on_lose_focus = window2_on_lose_focus_handler
+
+    app.current_window = window1
+    window1_on_gain_focus_handler.assert_called_once_with(window1)
```

Comment:
These calls assertions will be potentially misleading, as the mocks haven't been reset between each use. The test also isn't asserting the handlers that *aren't* called.

```suggestion
    window1_on_gain_focus_handler.assert_called_once_with(window1)
    window1_on_lose_focus_handler.assert_not_called()
    window2_on_gain_focus_handler.assert_not_called()
    window2_on_lose_focus_handler.assert_not_called()

    window1_on_gain_focus_handler.reset_mock()
    window1_on_lose_focus_handler.reset_mock()
    window2_on_gain_focus_handler.reset_mock()
    window2_on_lose_focus_handler.reset_mock()
```

Given that this will be a repeated pattern, it might be worth wrapping this sequence of calls in a utility assert methods.